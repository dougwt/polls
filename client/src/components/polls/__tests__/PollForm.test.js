import React from 'react';
import { shallow, mount } from 'enzyme';
import { Provider } from 'react-redux';
import { reducer as formReducer } from 'redux-form';
import { createStore, combineReducers, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { StaticRouter } from 'react-router';
import ReduxForm, { PollForm } from '../PollForm';

let props;
let wrapper;

describe('PollForm', () => {
  beforeEach(() => {
    props = {
      onPollSubmit: jest.fn(),
      handleSubmit: jest.fn()
    };

    wrapper = shallow(<PollForm {...props} />);
  });

  it('renders properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('shows a Form component', () => {
    expect(wrapper.find('form').exists()).toBe(true);
  });

  // TODO: Add owner field to form as a hidden field
  // xit('shows a hidden `Owner` field', () => {
  //   expect(wrapper.find('input[name="owner"]').exists()).toBe(true);
  // });

  it('should ask for a question', () => {
    expect(wrapper.find('Field[name="question"]').exists()).toBe(true);
  });

  it('should ask for choices', () => {
    expect(wrapper.find('FieldArray[name="choices"]').exists()).toBe(true);
  });

  it('shows a `Preview` button', () => {
    expect(wrapper.find('.btn-preview').exists()).toBe(true);
  });

  it('shows a `Cancel` button', () => {
    expect(wrapper.find('.btn-cancel').exists()).toBe(true);
  });

  describe('when the `Preview` button is clicked', () => {
    let store;

    beforeEach(() => {
      store = createStore(
        combineReducers({ form: formReducer }),
        applyMiddleware(thunk)
      );
      const initialValues = { choices: [0, 1] };

      wrapper = mount(
        <Provider store={store}>
          <StaticRouter context={{}}>
            <ReduxForm {...props} initialValues={initialValues} />
          </StaticRouter>
        </Provider>
      );
    });

    it('submits the form', () => {
      wrapper.find('Button.btn-preview').simulate('click');
      expect(props.handleSubmit).toHaveBeenCalled();
    });

    describe('when the form validation fails', () => {
      beforeEach(() => {
        wrapper.find('input[name="question"]').simulate('change', {
          target: { value: 'Would you like to play a game?' }
        });
        wrapper.find('Button.btn-preview').simulate('click');
        expect(props.handleSubmit).toHaveBeenCalled();
      });

      it('shows an error', () => {
        const form = store.getState().form.pollForm;
        expect(form.values.question).toBe('Would you like to play a game?');
        expect(form.syncErrors.choice_1).toBe(
          'You must provide text for Choice 1'
        );
        expect(form.syncErrors.choice_2).toBe(
          'You must provide text for Choice 2'
        );
      });
    });

    describe('when the form validation succeeds', () => {
      beforeEach(() => {
        wrapper.find('input[name="question"]').simulate('change', {
          target: { value: 'Would you like to play a game?' }
        });
        wrapper
          .find('input[name="choice_1"]')
          .simulate('change', { target: { value: 'Yes' } });
        wrapper
          .find('input[name="choice_2"]')
          .simulate('change', { target: { value: 'No' } });
        wrapper.find('Button.btn-preview').simulate('click');
        expect(props.handleSubmit).toHaveBeenCalled();
      });

      // TODO: Add a test to confirm user is shown the next component
      it('does not show any errors', () => {
        const form = store.getState().form.pollForm;
        expect(form.values.question).toBe('Would you like to play a game?');
        expect(form.values.choice_1).toBe('Yes');
        expect(form.values.choice_2).toBe('No');
        expect(form.syncErrors).toBeUndefined();
      });
    });
  });

  describe('when the `Cancel` button is clicked', () => {
    it('returns to the Dashboard when Back button is clicked', () => {
      expect(wrapper.find('.btn-cancel').props().to).toEqual('/polls');
    });
  });

  describe('renderChoices', () => {
    let store;

    beforeEach(() => {
      store = createStore(combineReducers({ form: formReducer }));
    });

    describe('when there are 2 choices', () => {
      beforeEach(() => {
        const initialValues = { choices: [0, 1] };

        wrapper = mount(
          <Provider store={store}>
            <StaticRouter context={{}}>
              <ReduxForm {...props} initialValues={initialValues} />
            </StaticRouter>
          </Provider>
        );
      });

      it('shows an enabled `Add Choice` button', () => {
        expect(wrapper.find('Button.btn-add-choice').exists()).toBe(true);
        expect(wrapper.find('Button.btn-add-choice').hasClass('disabled')).toBe(
          false
        );
      });

      it('allows the user to add an additional choice', () => {
        expect(store.getState().form.pollForm.values.choices.length).toEqual(2);
        wrapper.find('Button.btn-add-choice').simulate('click');
        expect(store.getState().form.pollForm.values.choices.length).toEqual(3);
      });

      it('shows a disabled `Remove Choice` button', () => {
        expect(wrapper.find('Button.btn-remove-choice').exists()).toBe(true);
        expect(
          wrapper.find('Button.btn-remove-choice').hasClass('disabled')
        ).toBe(true);
      });

      it('prevents the user from removing an existing choice', () => {
        expect(store.getState().form.pollForm.values.choices.length).toEqual(2);
        wrapper.find('Button.btn-remove-choice').simulate('click');
        expect(store.getState().form.pollForm.values.choices.length).toEqual(2);
      });
    });

    describe('when there are 3 choices', () => {
      beforeEach(() => {
        const initialValues = { choices: [0, 1, 2] };

        wrapper = mount(
          <Provider store={store}>
            <StaticRouter context={{}}>
              <ReduxForm {...props} initialValues={initialValues} />
            </StaticRouter>
          </Provider>
        );
      });

      it('shows an enabled `Add Choice` button', () => {
        expect(wrapper.find('Button.btn-add-choice').exists()).toBe(true);
        expect(wrapper.find('Button.btn-add-choice').hasClass('disabled')).toBe(
          false
        );
      });

      it('allows the user to add an additional choice', () => {
        expect(store.getState().form.pollForm.values.choices.length).toEqual(3);
        wrapper.find('Button.btn-add-choice').simulate('click');
        expect(store.getState().form.pollForm.values.choices.length).toEqual(4);
      });

      it('shows an enabled `Remove Choice` button', () => {
        expect(wrapper.find('Button.btn-remove-choice').exists()).toBe(true);
        expect(
          wrapper.find('Button.btn-remove-choice').hasClass('disabled')
        ).toBe(false);
      });

      it('allows the user to remove an existing choice', () => {
        expect(store.getState().form.pollForm.values.choices.length).toEqual(3);
        wrapper.find('Button.btn-remove-choice').simulate('click');
        expect(store.getState().form.pollForm.values.choices.length).toEqual(2);
      });
    });

    describe('when there are 4 choices', () => {
      beforeEach(() => {
        const initialValues = { choices: [0, 1, 2, 3] };

        wrapper = mount(
          <Provider store={store}>
            <StaticRouter context={{}}>
              <ReduxForm {...props} initialValues={initialValues} />
            </StaticRouter>
          </Provider>
        );
      });

      it('shows an enabled `Add Choice` button', () => {
        expect(wrapper.find('Button.btn-add-choice').exists()).toBe(true);
        expect(wrapper.find('Button.btn-add-choice').hasClass('disabled')).toBe(
          false
        );
      });

      it('allows the user to add an additional choice', () => {
        expect(store.getState().form.pollForm.values.choices.length).toEqual(4);
        wrapper.find('Button.btn-add-choice').simulate('click');
        expect(store.getState().form.pollForm.values.choices.length).toEqual(5);
      });

      it('shows an enabled `Remove Choice` button', () => {
        expect(wrapper.find('Button.btn-remove-choice').exists()).toBe(true);
        expect(
          wrapper.find('Button.btn-remove-choice').hasClass('disabled')
        ).toBe(false);
      });

      it('allows the user to remove an existing choice', () => {
        expect(store.getState().form.pollForm.values.choices.length).toEqual(4);
        wrapper.find('Button.btn-remove-choice').simulate('click');
        expect(store.getState().form.pollForm.values.choices.length).toEqual(3);
      });
    });

    describe('when there are 5 choices', () => {
      beforeEach(() => {
        const initialValues = { choices: [0, 1, 2, 3, 4] };

        wrapper = mount(
          <Provider store={store}>
            <StaticRouter context={{}}>
              <ReduxForm {...props} initialValues={initialValues} />
            </StaticRouter>
          </Provider>
        );
      });

      it('shows an enabled `Add Choice` button', () => {
        expect(wrapper.find('Button.btn-add-choice').exists()).toBe(true);
        expect(wrapper.find('Button.btn-add-choice').hasClass('disabled')).toBe(
          false
        );
      });

      it('allows the user to add an additional choice', () => {
        expect(store.getState().form.pollForm.values.choices.length).toEqual(5);
        wrapper.find('Button.btn-add-choice').simulate('click');
        expect(store.getState().form.pollForm.values.choices.length).toEqual(6);
      });

      it('shows an enabled `Remove Choice` button', () => {
        expect(wrapper.find('Button.btn-remove-choice').exists()).toBe(true);
        expect(
          wrapper.find('Button.btn-remove-choice').hasClass('disabled')
        ).toBe(false);
      });

      it('allows the user to remove an existing choice', () => {
        expect(store.getState().form.pollForm.values.choices.length).toEqual(5);
        wrapper.find('Button.btn-remove-choice').simulate('click');
        expect(store.getState().form.pollForm.values.choices.length).toEqual(4);
      });
    });

    describe('when there are 6 choices', () => {
      beforeEach(() => {
        const initialValues = { choices: [0, 1, 2, 3, 4, 5] };

        wrapper = mount(
          <Provider store={store}>
            <StaticRouter context={{}}>
              <ReduxForm {...props} initialValues={initialValues} />
            </StaticRouter>
          </Provider>
        );
      });

      it('shows a disabled `Add Choice` button', () => {
        expect(wrapper.find('Button.btn-add-choice').exists()).toBe(true);
        expect(wrapper.find('Button.btn-add-choice').hasClass('disabled')).toBe(
          true
        );
      });

      it('prevents the user from adding an additional choice', () => {
        expect(store.getState().form.pollForm.values.choices.length).toEqual(6);
        wrapper.find('Button.btn-add-choice').simulate('click');
        expect(store.getState().form.pollForm.values.choices.length).toEqual(6);
      });

      it('shows an enabled `Remove Choice` button', () => {
        expect(wrapper.find('Button.btn-remove-choice').exists()).toBe(true);
        expect(
          wrapper.find('Button.btn-remove-choice').hasClass('disabled')
        ).toBe(false);
      });

      it('allows the user to remove an existing choice', () => {
        expect(store.getState().form.pollForm.values.choices.length).toEqual(6);
        wrapper.find('Button.btn-remove-choice').simulate('click');
        expect(store.getState().form.pollForm.values.choices.length).toEqual(5);
      });
    });
  });
});
