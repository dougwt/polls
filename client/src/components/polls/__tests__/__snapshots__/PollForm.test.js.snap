// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PollForm renders properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    error={null}
    handleSubmit={[Function]}
    onCancel={[Function]}
    onSubmit={[Function]}
    waiting={false}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <form
        onSubmit={undefined}
>
        <Row>
                <Field
                        component={[Function]}
                        icon="question_answer"
                        label="What question would you like to ask?"
                        name="question"
                />
        </Row>
        <Row>
                <FieldArray
                        component={[Function]}
                        name="choices"
                />
        </Row>
        <Row>
                 
        </Row>
        <Row>
                <Button
                        className="btn btn-back red white-text"
                        node="button"
                        onClick={[Function]}
                        type="button"
                >
                        Cancel
                        <Icon
                                left={true}
                        >
                                close
                        </Icon>
                </Button>
                <Button
                        className="btn-next teal right white-text"
                        node="button"
                        waves="light"
                >
                        Next
                        <Icon
                                right={true}
                        >
                                keyboard_arrow_right
                        </Icon>
                </Button>
        </Row>
</form>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Row>
            <Field
                        component={[Function]}
                        icon="question_answer"
                        label="What question would you like to ask?"
                        name="question"
            />
</Row>,
          <Row>
            <FieldArray
                        component={[Function]}
                        name="choices"
            />
</Row>,
          <Row>
             
</Row>,
          <Row>
            <Button
                        className="btn btn-back red white-text"
                        node="button"
                        onClick={[Function]}
                        type="button"
            >
                        Cancel
                        <Icon
                                    left={true}
                        >
                                    close
                        </Icon>
            </Button>
            <Button
                        className="btn-next teal right white-text"
                        node="button"
                        waves="light"
            >
                        Next
                        <Icon
                                    right={true}
                        >
                                    keyboard_arrow_right
                        </Icon>
            </Button>
</Row>,
        ],
        "onSubmit": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <Field
              component={[Function]}
              icon="question_answer"
              label="What question would you like to ask?"
              name="question"
/>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "component": [Function],
              "icon": "question_answer",
              "label": "What question would you like to ask?",
              "name": "question",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <FieldArray
              component={[Function]}
              name="choices"
/>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "component": [Function],
              "name": "choices",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": " ",
          },
          "ref": null,
          "rendered": " ",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <Button
                className="btn btn-back red white-text"
                node="button"
                onClick={[Function]}
                type="button"
>
                Cancel
                <Icon
                                left={true}
                >
                                close
                </Icon>
</Button>,
              undefined,
              <Button
                className="btn-next teal right white-text"
                node="button"
                waves="light"
>
                Next
                <Icon
                                right={true}
                >
                                keyboard_arrow_right
                </Icon>
</Button>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  "Cancel",
                  <Icon
                    left={true}
>
                    close
</Icon>,
                ],
                "className": "btn btn-back red white-text",
                "node": "button",
                "onClick": [Function],
                "type": "button",
              },
              "ref": null,
              "rendered": Array [
                "Cancel",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "children": "close",
                    "left": true,
                  },
                  "ref": null,
                  "rendered": "close",
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
            undefined,
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  "Next",
                  <Icon
                    right={true}
>
                    keyboard_arrow_right
</Icon>,
                ],
                "className": "btn-next teal right white-text",
                "node": "button",
                "waves": "light",
              },
              "ref": null,
              "rendered": Array [
                "Next",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "children": "keyboard_arrow_right",
                    "right": true,
                  },
                  "ref": null,
                  "rendered": "keyboard_arrow_right",
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": "form",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <form
          onSubmit={undefined}
>
          <Row>
                    <Field
                              component={[Function]}
                              icon="question_answer"
                              label="What question would you like to ask?"
                              name="question"
                    />
          </Row>
          <Row>
                    <FieldArray
                              component={[Function]}
                              name="choices"
                    />
          </Row>
          <Row>
                     
          </Row>
          <Row>
                    <Button
                              className="btn btn-back red white-text"
                              node="button"
                              onClick={[Function]}
                              type="button"
                    >
                              Cancel
                              <Icon
                                        left={true}
                              >
                                        close
                              </Icon>
                    </Button>
                    <Button
                              className="btn-next teal right white-text"
                              node="button"
                              waves="light"
                    >
                              Next
                              <Icon
                                        right={true}
                              >
                                        keyboard_arrow_right
                              </Icon>
                    </Button>
          </Row>
</form>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Row>
              <Field
                            component={[Function]}
                            icon="question_answer"
                            label="What question would you like to ask?"
                            name="question"
              />
</Row>,
            <Row>
              <FieldArray
                            component={[Function]}
                            name="choices"
              />
</Row>,
            <Row>
               
</Row>,
            <Row>
              <Button
                            className="btn btn-back red white-text"
                            node="button"
                            onClick={[Function]}
                            type="button"
              >
                            Cancel
                            <Icon
                                          left={true}
                            >
                                          close
                            </Icon>
              </Button>
              <Button
                            className="btn-next teal right white-text"
                            node="button"
                            waves="light"
              >
                            Next
                            <Icon
                                          right={true}
                            >
                                          keyboard_arrow_right
                            </Icon>
              </Button>
</Row>,
          ],
          "onSubmit": undefined,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <Field
                component={[Function]}
                icon="question_answer"
                label="What question would you like to ask?"
                name="question"
/>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "component": [Function],
                "icon": "question_answer",
                "label": "What question would you like to ask?",
                "name": "question",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <FieldArray
                component={[Function]}
                name="choices"
/>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "component": [Function],
                "name": "choices",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": " ",
            },
            "ref": null,
            "rendered": " ",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <Button
                  className="btn btn-back red white-text"
                  node="button"
                  onClick={[Function]}
                  type="button"
>
                  Cancel
                  <Icon
                                    left={true}
                  >
                                    close
                  </Icon>
</Button>,
                undefined,
                <Button
                  className="btn-next teal right white-text"
                  node="button"
                  waves="light"
>
                  Next
                  <Icon
                                    right={true}
                  >
                                    keyboard_arrow_right
                  </Icon>
</Button>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": Array [
                    "Cancel",
                    <Icon
                      left={true}
>
                      close
</Icon>,
                  ],
                  "className": "btn btn-back red white-text",
                  "node": "button",
                  "onClick": [Function],
                  "type": "button",
                },
                "ref": null,
                "rendered": Array [
                  "Cancel",
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "children": "close",
                      "left": true,
                    },
                    "ref": null,
                    "rendered": "close",
                    "type": [Function],
                  },
                ],
                "type": [Function],
              },
              undefined,
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": Array [
                    "Next",
                    <Icon
                      right={true}
>
                      keyboard_arrow_right
</Icon>,
                  ],
                  "className": "btn-next teal right white-text",
                  "node": "button",
                  "waves": "light",
                },
                "ref": null,
                "rendered": Array [
                  "Next",
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "children": "keyboard_arrow_right",
                      "right": true,
                    },
                    "ref": null,
                    "rendered": "keyboard_arrow_right",
                    "type": [Function],
                  },
                ],
                "type": [Function],
              },
            ],
            "type": [Function],
          },
        ],
        "type": "form",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
